src/transports/http.h

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#ifndef INCLUDE_http_h__
+#define INCLUDE_http_h__
+
+#include "buffer.h"
+
+#define GIT_HTTP_USER_AGENT_FMT "git/2.0 (%s)"
```

Comment:
Considering we never use `git_http__user_agent` without `GIT_HTTP_USER_AGENT_FMT`, why even bother making it a macro instead of just letting `git_http__user_agent` assemble the complete string?

Reply:
It would be an awkward flow, since the fundamental strings are constructed differently for the two platforms.  One is complete, the other has `User-Agent: ` appended and newlines suffixed.

Reply:
Sure, I got that. The following easily handles that scenario but is still a lot more readable, I'd say:

```
diff --git a/src/transports/http.c b/src/transports/http.c
index 70a299322..6bebca910 100644
--- a/src/transports/http.c
+++ b/src/transports/http.c
@@ -202,8 +202,7 @@ static int gen_request(
 
 	git_buf_printf(buf, "%s %s%s HTTP/1.1\r\n", s->verb, path, s->service_url);
 
-	git_buf_printf(buf, "User-Agent: " GIT_HTTP_USER_AGENT_FMT "\r\n",
-		git_http__user_agent());
+	git_buf_printf(buf, "User-Agent: %s\r\n", git_http__user_agent());
 	git_buf_printf(buf, "Host: %s\r\n", t->connection_data.host);
 
 	if (s->chunked || content_length > 0) {
diff --git a/src/transports/http.h b/src/transports/http.h
index 0ad1f80bf..f34ef8ee9 100644
--- a/src/transports/http.h
+++ b/src/transports/http.h
@@ -10,8 +10,6 @@
 
 #include "buffer.h"
 
-#define GIT_HTTP_USER_AGENT_FMT "git/2.0 (%s)"
-
 GIT_INLINE(const char *) git_http__user_agent(void)
 {
 	const char *custom = git_libgit2__user_agent();
@@ -19,7 +17,7 @@ GIT_INLINE(const char *) git_http__user_agent(void)
 	if (custom)
 		return custom;
 
-	return "libgit2 " LIBGIT2_VERSION;
+	return "git/2.0 (libgit2 "LIBGIT2_VERSION")";
 }
 
 #endif
diff --git a/src/transports/winhttp.c b/src/transports/winhttp.c
index c004dccc0..5138e3993 100644
--- a/src/transports/winhttp.c
+++ b/src/transports/winhttp.c
@@ -758,7 +758,7 @@ static int winhttp_connect(
 		return -1;
 	}
 
-	if ((error = git_buf_printf(&ua, GIT_HTTP_USER_AGENT_FMT, git_http__user_agent())) < 0) {
+	if ((error = git_buf_puts(&ua, git_http__user_agent())) < 0) {
 		git__free(wide_host);
 		return error;
 	}
```

I think the _real_ issue here is the custom user agent, though. We _could_ do something like the following to also enable the above with custom user agents. But I'm not sure I like that.

```
diff --git a/src/settings.c b/src/settings.c
index 2a52ffbf6..0fa6df13b 100644
--- a/src/settings.c
+++ b/src/settings.c
@@ -181,8 +181,12 @@ int git_libgit2_opts(int key, ...)
 #endif
 		break;
 	case GIT_OPT_SET_USER_AGENT:
+	{
+		git_buf ua = GIT_BUF_INIT;
+		git_buf_printf(&ua, "git/2.0 (%s)", va_arg(ap, const char *));
+
 		git__free(git__user_agent);
-		git__user_agent = git__strdup(va_arg(ap, const char *));
+		git__user_agent = git_buf_detach(&ua);
 		if (!git__user_agent) {
 			giterr_set_oom();
 			error = -1;
@@ -190,6 +194,7 @@ int git_libgit2_opts(int key, ...)
 
 		break;
 
+	}
 	case GIT_OPT_ENABLE_STRICT_OBJECT_CREATION:
 		git_object__strict_input_validation = (va_arg(ap, int) != 0);
 		break;
```

Reply:
Right - your pattern has the same problem that we have now:  we're going to forget to change both places, and we'll end custom user-agents as `git/2.0 (my custom user agent)` and our default user agents as `git/3.0 (libgit2 ...)`.

Reply:
I agree. Seems like the _real_ issue is the API here. If we were able to return an allocated string we could just move assembling the custom user agent string into `git_http__user_agent`, as well, and be done about it.

Reply:
There's no need to introduce another allocation here.